### PR TITLE
Make DynamicPool allocation behavior consistent with previous release

### PR DESCRIPTION
**Pull request by jdahm**

### CODE REVIEW

File: tests/unit/util/CMakeLists.txt [hunk 1]
```diff
@@ -16,10 +16,24 @@ blt_add_test(
   NAME exception_tests
   COMMAND exception_tests)
 
+set (allocation_map_tests_depends
+  umpire gtest)
+
+if (ENABLE_CUDA)
+  set (allocation_map_tests_depends
```
Comment (mcfadden8) [inline tests/unit/util/CMakeLists.txt]: Was this a regression?  How did this new dependency come about and how is this related to this specific PR?

File: src/umpire/strategy/DynamicPool.cpp [hunk 2]
```diff
@@ -15,6 +15,7 @@
 #include <cstdlib>
 #include <algorithm>
 #include <sstream>
+#include <new>
```
Comment (mcfadden8) [inline src/umpire/strategy/DynamicPool.cpp]: Was this added to get std::bad_alloc{} throw?

File: src/umpire/strategy/DynamicPool.cpp [hunk 3]
```diff
@@ -114,21 +115,56 @@ DynamicPool::findFreeBlock(std::size_t bytes) const
...
+    }
+    catch (...) {
+      UMPIRE_LOG(Error,
+                 "\n\tUnable to allocate from resource even after giving back free blocks.\n"
+                 "\tThrowing to let application know we have no more memory: "
+                 << getCurrentSize() << " Bytes still allocated to pool\n"
+                 << getFreeBlocks() << " Partially Free Blocks, "
+                 << getInUseBlocks() << " Used Blocks\n"
+        );
+      throw std::bad_alloc{};
```
Comment (mcfadden8) [inline src/umpire/strategy/DynamicPool.cpp]: I'm not sure about changing the exception value here.  The previous implementation was intending to perform a re-throw of the previously thrown exception (by calling throw with no arguments).

File: tests/unit/util/CMakeLists.txt (hunk 1 above)
Comment (jdahm) [reply tests/unit/util/CMakeLists.txt]: I don't know. I saw a HIP build error on the previous commit and added it. BLT isn't carrying over the dependencies from umpire correctly, because e.g. `cuda` implies `cuda_runtime`. It's complicated...

File: src/umpire/strategy/DynamicPool.cpp (hunk 2 above)
Comment (jdahm) [reply src/umpire/strategy/DynamicPool.cpp]: Yes.

File: src/umpire/strategy/DynamicPool.cpp (hunk 3 above)
Comment (jdahm) [reply src/umpire/strategy/DynamicPool.cpp]: I didn't know that throw without args threw the previous exception. Should I switch it back to that?

Comment (davidbeckingsale) [reply src/umpire/strategy/DynamicPool.cpp]: Please switch it back - our exceptions contain additonal info that bad_alloc won't, e.g. CUDA error converted to its string description.

File: src/umpire/strategy/DynamicPool.cpp (hunk 2 above)
Comment (davidbeckingsale) [reply src/umpire/strategy/DynamicPool.cpp]: Please remove this.

File: src/umpire/strategy/DynamicPool.cpp (hunk 3 above)
Comment (mcfadden8) [reply src/umpire/strategy/DynamicPool.cpp]: Yes, please switch it back so that umpire exception information information is re-thrown.
